### PR TITLE
Refs Task#382209 #376320 Unit test list, create webhooks

### PR DESCRIPTION
**Pull request by quanghung97**
## Related Tickets

- [#382209](https://dev.sun-asterisk.com/issues/382209)
- [#376320](https://dev.sun-asterisk.com/issues/376320)

## What's this PR do ?

- [x] Unit test list, create Webhooks
- [x] Refactor Repositories

## Performance

- [ ] Resolved n + 1 query
- [ ] Run explain query already
- [ ] Time run rake task : 1000 ms

## Checklist

- [ ] It was tested in local success?
- [x] Fill link PR into ticket and the opposite
- [ ] Note purpose/reason, solution, scope of influence into ticket

## ENV note

```
```

## Notes
*(Other notes)*

### CODE REVIEW

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.15</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.13</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.15</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.13</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.9</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.15</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>composer_dump-autolo._start_run_.13</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.17</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_cache_cl._start_run_.9</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.17</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_cache_cl._start_run_.9</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.17</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_cache_cl._start_run_.9</td><td>success</td><td></td></tr></table>

**Comment by framgiaciv3** [comment]
coverage</br> **1. Result build** </br><table><tr><th>Style</th><th>Status</th><th>Ignored</th></tr><tr><td>composer_dump-autolo._start_run_.15</td><td>success</td><td></td></tr><tr><td>phpcs</td><td>success</td><td></td></tr><tr><td>php_artisan_key_gene._start_run_.11</td><td>success</td><td></td></tr><tr><td>composer_install._start_run_.5</td><td>success</td><td></td></tr><tr><td>php_artisan_config_c._start_run_.7</td><td>success</td><td></td></tr><tr><td>cp__env_example__env._start_run_.3</td><td>success</td><td></td></tr><tr><td>php_artisan_migrate_._start_run_.17</td><td>success</td><td></td></tr><tr><td>framgia-ci_test-conn._start_run_.13</td><td>success</td><td></td></tr><tr><td>phpunit</td><td>failure</td><td></td></tr><tr><td>php_artisan_cache_cl._start_run_.9</td><td>success</td><td></td></tr></table>

File: tests/Unit/Repositories/WebhookRepositoryTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        // mock instance Repository
+        $mock = Mockery::mock('App\Repositories\Interfaces\WebhookRepositoryInterface');
+
+        $mock->shouldReceive('create');
+
+        Auth::shouldReceive('id')->once()->andReturn(1);
+
+        $this->app->instance('App\Repositories\Interfaces\WebhookRepositoryInterface', $mock);
+
+        $this->post('/webhooks', ['name' => 'string', 'description' => 'some thing successfully', 'bot_id' => 1, 'room_name' => 'string', 'room_id' => 1]);
```
Comment (ghost) [inline tests/Unit/Repositories/WebhookRepositoryTest.php]: ko thấy expect value trong đoạn này nhỉ

File: tests/Feature/WebhookControllerTest.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        $webhookLists = factory(Webhook::class, 2)->create();
+        factory(Bot::class, 2)->create();
+        $user = $webhookLists[0]->user;
+
+        $this->actingAs($user);
+        $response = $this->post(route('webhooks.store'), ['name' => 'string', 'description' => 'some thing successfully', 'bot_id' => 1, 'room_name' => 'string', 'room_id' => 1]);
+
+        $response->assertRedirect();
+    }
```
Comment (ghost) [inline tests/Feature/WebhookControllerTest.php:66]: bổ sung thêm một số trường hợp failed như chưa login, fail validate đi cậu :D

File: tests/Unit/Repositories/WebhookRepositoryTest.php (hunk 1 above)
Comment (quanghung97) [reply tests/Unit/Repositories/WebhookRepositoryTest.php]: đoạn này có assert called time của create() và Auth::id() ở phía trên trước khi gọi route rồi.
nếu viết rõ hơn thì vất cái called times xuống dưới

**Review by nguyenthinhhn — APPROVED** [review]
ok

**Review by ghost — APPROVED** [review]

**Comment by phucqd-1305** [comment]
OK

**Pull request merged**